### DISCUSSION TITLE
Configure relationship to join on subquery

### DISCUSSION THREAD

**Discussion by lovetox** [sqlalchemy/sqlalchemy#11061] [Usage Questions] [answered]
Hi,

I invested some hours in finding examples and reading docs, but everything i could find are very basic where conditions on joins

I use sqlite as database.

I have one table like this

```python
class Message(MappedAsDataclass, Base, kw_only=True):
    __tablename__ = "message"

    pk: Mapped[int] = mapped_column(primary_key=True, init=False)

    account: Mapped[str]
    occupant_id: Mapped[int | None]
    resource: Mapped[str]
    type: Mapped[int]

    message_id: Mapped[str]
    message: Mapped[str]

    correction_id: Mapped[str | None]     # Can be set to a message_id

    correction: Mapped[list[Message]] = relationship(....)
```

The Message table contains messages but some messages are corrections of other messages. This is indicated by the fields `correction_id` set to a `message_id`, which means this message is a correction of the message with `message_id`. All corrections need to reference the original first message, so there is no linked list where every next correction references the correction before, if thats of relevance.

There can be multiple corrections, and i want all of them in a list in the attribute `correction`.

The comparisons in the subquery are a bit complicated and look like this

```sql
message.message_id = message1.correction_id AND
message.account = message1.account AND
message.occupant_id IS message1.occupant_id AND
CASE WHEN message.type = 2 AND message.occupant_id IS NULL THEN
     message.resource = message1.resource
ELSE 1
END
```

Some things that are important

1. occupant_id can be NULL, i need the "IS" operator
2. I seem to need the case statement, not sure if this even works in a join

I tried to do it without subquery and simply specifying all these conditions in `primaryjoin`, but i failed to make the "IS" operator work, i didn't even come to try the CASE stuff.

Then i thought i can simply do this as subquery and join the result, but failed to find any example where a subquery is used in a relationship.

Any help is very much appreciated.

Thanks

**Comment by CaselIT** [comment, marked as answer]
Hi,

Not an easy condition but all this is explained here https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#creating-custom-foreign-conditions

In particular this is the running example:
```py
from __future__ import annotations

from sqlalchemy import and_
from sqlalchemy import case
from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session
from sqlalchemy.orm import foreign
from sqlalchemy.orm import remote
from sqlalchemy.orm.strategy_options import joinedload


class Base(DeclarativeBase):
    pass


class Message(Base):
    __tablename__ = "message"

    pk: Mapped[int] = mapped_column(primary_key=True)

    account: Mapped[str] = mapped_column()
    occupant_id: Mapped[int | None] = mapped_column()
    resource: Mapped[str]
    type: Mapped[int] = mapped_column()

    message_id: Mapped[str] = mapped_column()
    message: Mapped[str]

    correction_id: Mapped[str | None]

    correction: Mapped[list[Message]] = relationship(
        primaryjoin=and_(
            foreign(message_id) == remote(message_id),
            foreign(account) == remote(account),
            foreign(occupant_id).is_(remote(occupant_id)),
            case(
                (
                    and_(
                        foreign(type) == 2,
                        foreign(occupant_id).is_(None),
                    ),
                    None,
                ),
                else_=1,
            ),
        ),
        viewonly=True,
    )


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)
stmt = select(Message).options(joinedload(Message.correction))
with Session(e) as s:
    s.execute(stmt).all()
```

this prints:
```sql
SELECT message.pk, message.account, message.occupant_id, message.resource, message.type, message.message_id, message.message, message.correction_id, message_1.pk AS pk_1, message_1.account AS account_1, message_1.occupant_id AS occupant_id_1, message_1.resource AS resource_1, message_1.type AS type_1, message_1.message_id AS message_id_1, message_1.message AS message_2, message_1.correction_id AS correction_id_1
FROM message
LEFT OUTER JOIN message AS message_1 ON 
  message.message_id = message_1.message_id AND
  message.account = message_1.account AND
  message.occupant_id IS message_1.occupant_id AND
  CASE WHEN (message.type = ? AND message.occupant_id IS NULL) THEN NULL ELSE ? END
2024-02-27 21:25:34,468 INFO sqlalchemy.engine.Engine [generated in 0.00019s] (2, 1)
```

Note that you need to add ` = mapped_column()` when using column inline with the class declaration, since otherwise python does not define an annotation only class attribute

**Reply by lovetox** [reply]
the mapped_column() did trip me, thats why nothing worked, and then i tried to define it as a string ..

This works for me, thank you very much.

**Reply by CaselIT** [reply]
Yes that may be confusing but that's how python works. Sqlalchemy will create the columns for you but that's in the `__init_subclass__` method that runs after the class declaration
